### PR TITLE
Fixed  audio and MIDI drivers; encode device names as CP_UTF8

### PR DESCRIPTION
**Pull request by pedrolcl**
This also sets unicode (utf8) for CLI program's console output. It should be the proper fix for #983 wasapi problems.
cc: @tsingakbar 

Solves #1322

### CODE REVIEW

**Comment by derselbst** [comment]
Thanks Pedro! @tsingakbar If you have some time, could you pls. run `fluidsynth.exe -Q` and let us know if the reported WASAPI device names still look correct? For convenience, you may use the `fluidsynth-x64` prebuilt binaries: https://dev.azure.com/tommbrt/tommbrt/_build/results?buildId=10122&view=artifacts&pathAsName=false&type=publishedArtifacts

**Comment by pedrolcl** [comment]
CC  @nanitaro , please test as well

**Comment by nanitaro** [comment]
> CC @nanitaro , please test as well

I ran fluidsynth.exe -Q on a test version (fluidsynth-x64).
Japanese information was output without problems as in 2.3.5 (official release version).
All variants (x64,mingw-x64,x86) have been checked.
I will upload the output log just to be sure.
[fluidsynth.txt](https://github.com/FluidSynth/fluidsynth/files/15163562/fluidsynth.txt)

File: src/drivers/fluid_dsound.c [hunk 1]
```diff
@@ -137,6 +155,7 @@ fluid_dsound_enum_callback2(LPGUID guid, LPCTSTR description, LPCTSTR module, LP
             return FALSE;
```
Comment (derselbst) [inline src/drivers/fluid_dsound.c:155]: `name` will be leaked when returning false here.

File: include/fluidsynth/settings.h [hunk 2]
```diff
@@ -48,6 +48,9 @@ extern "C" {
  *       // ...
  *     }
  * @endcode
+ * All string settings are encoded in UTF-8. This includes the names
+ * of the audio and MIDI devices, exposed as setting options.
```
Comment (derselbst) [inline include/fluidsynth/settings.h:52]: Probably it's also worth mentioning in the `fluidsettings.xml`, that starting with 2.3.6 all device names are expected to be UTF8 encoded.

File: src/drivers/fluid_dsound.c [hunk 3]
```diff
@@ -107,8 +107,18 @@ typedef struct
 BOOL CALLBACK
 fluid_dsound_enum_callback(LPGUID guid, LPCTSTR description, LPCTSTR module, LPVOID context)
 {
-    fluid_settings_t *settings = (fluid_settings_t *) context;
-    fluid_settings_add_option(settings, "audio.dsound.device", (const char *)description);
+    fluid_settings_t *settings = (fluid_settings_t *)context;
+    char *name;
+#if _UNICODE
+    int nsz = WideCharToMultiByte(CP_UTF8, 0, description, -1, 0, 0, 0, 0);
+    name = FLUID_ARRAY(char, nsz + 1);
```
Comment (derselbst) [inline src/drivers/fluid_dsound.c:114]: Is there a reason for allocating +1 byte? From my understanding, when `cchWideChar` is set to -1, the value returned by `WideCharToMultiByte` will always include the NULL byte. So technically, the very last byte of  `name` would be uninitialized. Not sure, if that's intended?

**Review by derselbst — COMMENTED** [review]
Nice, thank you! A few minor comments.

File: src/drivers/fluid_dsound.c (hunk 1 above)
Comment (pedrolcl) [reply src/drivers/fluid_dsound.c]: fixed

File: include/fluidsynth/settings.h (hunk 2 above)
Comment (pedrolcl) [reply include/fluidsynth/settings.h:52]: done

File: src/drivers/fluid_dsound.c (hunk 3 above)
Comment (pedrolcl) [reply src/drivers/fluid_dsound.c]: This was the result of copy/paste. But you are right: that is what the documentation states, so there should be no need for this precaution.

All drivers fixed.

**Comment by pedrolcl** [comment]
With these changes, the users may enjoy utf8 output and input when using cmd.exe (or other shells) in Windows. The input is relevant for Fluidsynth's own shell.  And this would be enough for libfluidsynth.dll clients, like Qsynth.

But there is still a loose thread: the CLI program's command line arguments are only processed/converted into utf8 for file names, but not other arguments. For instance, this would not work:

C:> fluidsynth.exe -o audio.wasapi.device=GroßeStraßenlautsprecher

**Comment by derselbst** [comment]
Sorry for the late response, busy schedule.

It looks good to me now, thanks! If you want and have some time feel free to also UTF8 convert the commandline args.

@nanitaro Would you mind running `fluidsynth.exe -Q` again with these binaries? @tsingakbar CC
https://dev.azure.com/tommbrt/tommbrt/_build/results?buildId=10135&view=artifacts&pathAsName=false&type=publishedArtifacts

If everything goes well, I would apply the commits onto the `release/2.3` and then merge it back to master. But probably not before next weekend, though.

**Comment by nanitaro** [comment]
@derselbst
I ran the new test binary.
I think there is no problem about outputting Japanese information.
[fluidsynth_05112024.txt](https://github.com/FluidSynth/fluidsynth/files/15279719/fluidsynth_05112024.txt)

Compared to the previous binary, the output result of the default device in shared and exclusive mode seems to have changed.
This time the result seems to be more accurate.

For your information, the performance and settings of the sound devices in the test environment are as follows
Default device = SPDIF interface (FX-AUDIO-DAC-X6)
Windows sound properties. Default format=2ch,24bit,96000Hz

**Comment by pedrolcl** [comment]
After the commits on May 1st, not only WASAPI but also the other Windows drivers are now updated to use UTF8. There is no equivalent of `-Q` for the other drivers, but it should be possible to use the shell, starting fluidsynth.exe without the argument and...

```
> settings
[...]
> info midi.driver
Options: winmidi
> info midi.winmidi.device
[...]
> info audio.driver
Options: dsound, waveout, wasapi
[...]
>  info audio.dsound.device
>  info audio.waveout.device
>  info audio.wasapi.device
```

The output of all these commands should be encoded now in UTF8.

**Review by derselbst — APPROVED** [review]
I have just applied the commit to the `release/2.3` branch and merged it to master. Thanks @pedrolcl  for the fix and @nanitaro for the tests!

**Pull request closed without merging**